src/gui/properties/speedplotview.cpp

```diff
@@ -118,55 +107,75 @@ namespace
 {
     // Accumulator overflow will be hit in worst case on longest used averaging span,
     // defined by divider value. Maximum divider is DIVIDER_24HOUR = 144
     // Using int32 for accumulator we get overflow when transfer speed reaches 2^31/144 ~~ 14.2 MBytes/s.
     // With quint64 this speed limit is 2^64/144 ~~ 114 PBytes/s.
     // This speed is inaccessible to an ordinary user.
-    m_accumulator.x += pointData.x;
+    ++m_counter;
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] += pointData.y[id];
-    m_counter = (m_counter + 1) % m_divider;
-    if (m_counter != 0)
-        return; // still accumulating
+        m_accumulator[id] += sampleData[id];
+
+    // system may go to sleep, that can cause very big elapsed interval
+    const milliseconds updateInterval {static_cast<int64_t>(BitTorrent::Session::instance()->refreshInterval() * 1.25)};
+    const milliseconds maxElapsed {std::max(updateInterval, m_resolution)};
+    const milliseconds elapsed {std::min(milliseconds {m_lastSampleTime.elapsed()}, maxElapsed)};
+    if (elapsed < m_resolution)
+        return false; // still accumulating
+
     // it is time final averaging calculations
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] /= m_divider;
-    m_accumulator.x /= m_divider;
+        m_accumulator[id] /= m_counter;
+
+    m_currentDuration += elapsed;
+
+    // remove extra data from front if we reached max duration
+    if (m_currentDuration > m_maxDuration)
+    {
+        // once we go above the max duration never go below that
+        // otherwise it will cause empty space in graphs
+        while((!m_sink.empty())
+               && ((m_currentDuration - m_sink.front().duration) > m_maxDuration))
+        {
+            m_currentDuration -= m_sink.front().duration;
+            m_sink.pop_front();
+        }
+    }
+
     // now flush out averaged data
-    m_sink.push_back(m_accumulator);
+    Q_ASSERT(m_sink.size() < m_sink.capacity());
+    m_sink.push_back({elapsed, m_accumulator});
+
+    // reset
     m_accumulator = {};
+    m_counter = 0;
+    m_lastSampleTime.restart();
+    return true;
 }
 
-bool SpeedPlotView::Averager::isReady() const
+const SpeedPlotView::DataCircularBuffer &SpeedPlotView::Averager::data() const
 {
-    return m_counter == 0;
+    return m_sink;
 }
 
 SpeedPlotView::SpeedPlotView(QWidget *parent)
     : QGraphicsView(parent)
-    , m_data5Min(MIN5_BUF_SIZE)
-    , m_data30Min(MIN30_BUF_SIZE)
-    , m_data6Hour(HOUR6_BUF_SIZE)
-    , m_data12Hour(HOUR12_BUF_SIZE)
-    , m_data24Hour(HOUR24_BUF_SIZE)
-    , m_currentData(&m_data5Min)
-    , m_averager30Min(DIVIDER_30MIN, m_data30Min)
-    , m_averager6Hour(DIVIDER_6HOUR, m_data6Hour)
-    , m_averager12Hour(DIVIDER_12HOUR, m_data12Hour)
-    , m_averager24Hour(DIVIDER_24HOUR, m_data24Hour)
-    , m_period(MIN5)
-    , m_viewablePointsCount(MIN5_SEC)
+    , m_averager5Min(5min, RESOLUTION_MIN5)
```

Comment:
Please use new style initialization:
```c++
m_averager5Min {5min, RESOLUTION_MIN5}
```

It's nothing hard to change one existing line too to keep it consistent.